### DISCUSSION TITLE
Unification of the NFS format

### DISCUSSION THREAD

**Discussion by poncovka** [rhinstaller/anaconda#4491] [Ideas]
## Summary

During the modularization of the additional repositories, I have noticed that we support two different NFS formats, `nfs:` and `nfs://`. We are trying to unify the support for the base repository and the additional repositories as much as possible to reduce the amount of code and prevent inconsistencies. Therefore, it would be great if we could unify the formats as well and eventually support only one of them.

**The benefits:**
* The Anaconda backend can start to use the unified format right away.
* It will improve the user experience, because the current state can be frustrating.
* Consistency, no special cases in the code and in the documentation.

**The plan:**
1. Start to support the unified NFS format everywhere.
2. The output kickstart file will use only the unified format.
3. Document the unified format and deprecate the previous one.
4. Show a warning for the deprecated format, but keep the support.
5. Eventually, show an error for the deprecated format and drop the support.

As you can see in the details below, it looks like the formats started to diverge with additional repositories, because only the `repo` kickstart command and the `inst.addrepo` boot option support the `nfs://` format. Based on that, it would make sense to choose the `nfs:` format. On the other side, this format can be confusing due to the optional `<options>` part that is for some reason at the beginning of the string instead of the end. Anyway, none of these formats follow [RFC 2224](https://www.rfc-editor.org/rfc/rfc2224.html) and we cannot use it without modifications, because it doesn't allow to specify mount options.

I propose to choose the `nfs:` format and deprecate the `nfs://` format. What do you think?

<details>

## The nfs:// format

**Scheme:**
`nfs://<server>:<path>`
`nfs://<server>`

**The inst.addrepo boot option** [[docs]](https://anaconda-installer.readthedocs.io/en/latest/boot-options.html#inst-addrepo)
`inst.addrepo=REPO_NAME,nfs://<server>:/<path>` 

**The repo kickstart command** [[docs]](https://pykickstart.readthedocs.io/en/latest/kickstart-docs.html#repo)
Fun fact! This is not documented in [RHEL 9](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html-single/performing_an_advanced_rhel_9_installation/index#repo_kickstart-commands-for-system-configuration) or [RHEL 8](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html-single/performing_an_advanced_rhel_8_installation/index#repo_kickstart-commands-for-system-configuration).
`repo  --name=REPO_NAME--baseurl=nfs://host:/path/to/repo`

## The nfs: format

**Scheme:**
`nfs:<options>:<server>:<path>`
`nfs:<server>:<path>`
`nfs:<server>`

**The inst.repo boot option** [[docs]](https://anaconda-installer.readthedocs.io/en/latest/boot-options.html#inst-repo)
`inst.repo=nfs:[<options>:]<server>:/<path>` 

**The inst.ks boot option** [[docs]](https://anaconda-installer.readthedocs.io/en/latest/boot-options.html#inst-ks)
`inst.ks=nfs:<next_server>:/<filename>`

**The driverdisk kickstart command** [[docs]](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html-single/performing_an_advanced_rhel_9_installation/index#driverdisk_kickstart-commands-for-installation-program-configuration-and-flow-control)
`driverdisk --source=nfs:host:/path/to/dd.img` 

## Other formats

**The nfs kickstart command** [[docs]](https://pykickstart.readthedocs.io/en/latest/kickstart-docs.html#nfs)
`nfs --server <hostname> --dir <directory> [--opts <options>]` 

</details>

**Comment by poncovka** [comment]
@jkonecny12 @jstodola @bcl What do you think about this proposal?

**Comment by jstodola** [comment]
> I propose to choose the nfs: format and deprecate the nfs:// format. What do you think?

I agree.

**Comment by poncovka** [comment]
Btw. dracut uses, for example, `root=nfs:[<server-ip>:]<root-dir>[:<nfs-options>]` [[docs]](https://man7.org/linux/man-pages/man7/dracut.cmdline.7.html). The interesting thing is that anaconda uses the dracut's [nfs_to_var](https://github.com/dracutdevs/dracut/blob/master/modules.d/95nfs/nfs-lib.sh#L12 ) function  for parsing nfs formats and this function supports all these styles:

* `nfs://<server>[:<port>]/<path>` defined by RFC2224
* `nfs:[<server-ip>:]<root-dir>[:<nfs-options>]` defined by dracut
* `nfs:<options>:<server>:/<path>` defined by anaconda

Therefore, it is possible that some of the anaconda's boot options can work with other variants of the nfs format and some [suspicions parts of the code](https://github.com/rhinstaller/anaconda/blob/master/dracut/fetch-driver-net.sh#L34) seem to confirm that even though it is not documented.

**Comment by bcl** [comment]
A quick look at the code history seems to imply that dracut and anaconda were adding this at the same time (late 2009) and IIRC at that time Anaconda wasn't using dracut yet, it was still using the loader written in C. So that may explain the different order of options :)

From pykickstart's perspective it doesn't care what the url is. It does have `nfs://` documented in the repo command but that is easily changed, and it doesn't mention options. Lorax and livemedia-creator also don't care, they use pykickstart and the parsed methods.

I would lean towards using the dracut format, with the options at the end, because that form is likely to be more familiar to users.

**Comment by poncovka** [comment]
I agree with @bcl, it would be better the support the dracut format with the mount options at the end. It shouldn't be too difficult:

* There will be no change for users who don't care about the mount options.
* Thanks to the `nfs_to_var` function, the amount of changes in the anaconda dracut module should be minimal or none.
* The installer has to be able to parse both formats and show a warning if the mount points are at the beginning of the string. We can use the `nfs_to_var` function for an inspiration and use a list of common nfs options to be more precise if necessary.

@jstodola @jkonecny12 Would it be acceptable?

**Reply by jstodola** [reply]
I would agree with Brian if we were implementing the nfs support in the installer from scratch.
But since the currently implemented support is documented and the users/customers use it, I don't like the idea of changing the order of the field in the nfs format. IMO, we should only do as few changes as necessary. Keep in mind that the new format would eventually have to be also supported by custom provisioning scripts and services (like Satellite?). I'm not convinced the benefit of moving `<options>` to the end of the string is worth it.

People using ` inst.repo=nfs:<options>:<server>:/<path>` are already more familiar with it :-)

**Reply by poncovka** [reply]
I feel like this change will eventually happen no matter what we decide here, because the mount options at the beginning make no sense. Making this change after the unification will be more difficult. Right now, the mount options are supported only by `inst.repo` and some other boot options. However, the unified format will be used everywhere and integrated into the DBus API.

**Reply by jkonecny12** [reply]
I'm more inclined to @jstodola . Using a Dracut format doesn't seems like a big benefit right now when people are already using our current formats. I think that it would be better for them to stick what they know.

However, true is that NFS solution from Dracut is taken from kernel: https://www.kernel.org/doc/Documentation/filesystems/nfs/nfsroot.txt which makes the format a bit more stable.

On one hand we shouldn't break API for users if we don't have a strong reason to so I'm opposed to change it to a completely new format but instead take this as a fix of inconsistency between API of Anaconda (fix the least used).
On the other hand the proposed format is used by Dracut and kernel which might simplify stuff in the future.

@jstodola given that the Dracut format is not Dracut invention, do you still think it is a good idea to go with the current way instead of switching to something which is more consistent in the Linux world?

**Reply by jstodola** [reply]
The dracut and kernel formats are not the same, there is no consistency in this world:
dracut: `nfs:[<server-ip>:]<root-dir>[:<nfs-options>]`
kernel: `nfsroot=[<server-ip>:]<root-dir>[,<nfs-options>]`

I'm not strongly against the dracut format, since the current behavior still has to be supported for a while.

The currently documented and the new format should be covered by automated tests to avoid possible regressions before the support for the current format is removed in RHEL-1x.

**Reply by jkonecny12** [reply]
Ha you are correct, they changed the syntax later in the Dracut:

https://github.com/dracutdevs/dracut/commit/5103130352ae19f69c6357ebcf4b600e835bf7d7

In that case I'm also against using the Dracut format. It seems that even they changed it and I'm not convinced that we should follow them now instead of kernel or another project. If none of these is taken as standard than migrating us to them gives no point.

I would be in favor of doing as minimal changes as possible which I think is `nfs:...` solution. It is also closer to the Dracut than the other we have.

**Reply by poncovka** [reply]
Ok, the `nfs:<options>:<server>:<path>` format seems to be the winner.
